---
title: "Gprime-pooled-continuous-analysis"
format: gfm
editor: visual
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(dpi=300,fig.width=7)
```

Load packages

```{r}
#install.packages("devtools")
#devtools::install_github("bmansfeld/QTLseqr")
# install.packages("vcfR")
#install.packages("SNPfiltR")
library(SNPfiltR)
library(QTLseqr)
library(ggplot2)
library(vcfR)
library(dplyr)
```

#### Load in dataset

```{r}
# read in vcf
DiNV_vcf <- read.vcfR("~/Desktop/KU/sequences/16Cq-DiNV-Test/freebayes/continuous-DiNV.vcf")

# filter vcf to remove sites (SNPs) with more than two alleles present (from SNPfiltR package)
DiNV_vcf <- filter_biallelic(DiNV_vcf)

# extract the number of reads supporting the reference allele for each sample at each SNP 
# DiNV_vcf has a column containing RO or reference allele information
# the extract.gt function will subset from that column
# the same column contains a lot of other information separated by : and , which is why a specific function is needed 
ref <- as.data.frame(extract.gt(DiNV_vcf, element = "RO"))
# extract the number of reads supporting the alternate allele for each sample at each SNP 
# AO is alternate allele
# because I removed all SNPs that are more than biallelic, there should only be 1 alternate allele
alt <- as.data.frame(extract.gt(DiNV_vcf, element = "AO"))
DiNV_SNPs<-as.data.frame(DiNV_vcf@fix[,c(1,2,4,5)])
# add the read counts supporting the reference and alternate alleles for each of the two sequenced strains as their own columns
DiNV_SNPs$inn.ref<-ref$INN
DiNV_SNPs$inn.alt<-alt$INN
DiNV_SNPs$vir.ref<-ref$VIR
DiNV_SNPs$vir.alt<-alt$VIR

# there might be some missing data in this file, which would come out as rows where either inn or vir have an NA for alleles 
# remove any rows that have an NA 
DiNV_SNPs <- na.omit(DiNV_SNPs)
```

#### Prepare SNP dataset for Gprime format. Based off of QTLSeqr program.

```{r}
# what I need is columns that say:
# chrom, position, ref base, alt base, and allele counts per sample
# but the allele counts have to be named AD_ALT.sample or AD_REF.sample 
# so I need AD_ALT.inn, AD_REF.inn, AD_ALT.vir, and AD_REF.vir

colnames(DiNV_SNPs)[5] ="AD_REF.inn"
colnames(DiNV_SNPs)[6] ="AD_ALT.inn"
colnames(DiNV_SNPs)[7] ="AD_REF.vir"
colnames(DiNV_SNPs)[8] ="AD_ALT.vir"

head(DiNV_SNPs)
# looks good so far

# save this as a csv because the program doesn't seem to want it in R 
write.csv(DiNV_SNPs, file = "~/Desktop/KU/sequences/16Cq-DiNV-Test/freebayes/DiNV_SNPs_For_Gprime.csv", row.names = FALSE)

```

#### More preparing the table and loading in

```{r}
# vignette says "We define the sample name for each of the bulks."
# from what I can tell, a bulk is a pooled sample or a bulked sample
# for whatever reason this program calls them a high bulk and a low bulk 
# but I'm not sure what those terms mean 

# naming the samples exactly what it says after AD_REF. 
# also want to 
HighBulk <- "inn"
LowBulk <- "vir"
# assuming this is the way they want to represent chromosomes
# they usually want you to specify which chromosomes to use
# because there is just one it doesn't matter probably 
Chroms <- "NC_040699.1"

# see if importing the table works 
Gprime_df <- importFromTable(file = "~/Desktop/KU/sequences/16Cq-DiNV-Test/freebayes/DiNV_SNPs_For_Gprime.csv",
                      highBulk = HighBulk,
                      lowBulk = LowBulk,
                      chromList = Chroms)

# for whatever reason it renamed my columns, so I will just have to remember that high is innubila cells and low is virilis cells
```

The import from table function has added in some columns. They mean:

-   GQ.HIGH - The genotype quality score, (how confident we are in the genotyping)
-   SNPindex.HIGH - The calculated SNP-index for the high bulk. SNP index is alternate allele depth over total read depth
-   Same as above for the low bulk
-   REF_FRQ - reference allele depth for both samples divided by total read depth for both samples
-   deltaSNP - The change in SNP index, high bulk index minus low bulk index

#### Looking at SNPs to get an idea for filtering

"QTLseqr offers some options for filtering that may help reduce noise and improve results. Filtering is mainly based on read depth for each SNP, such that we can try to eliminate SNPs with low confidence, due to low coverage, and SNPs that may be in repetitive regions and thus have inflated read depth. You can also filter based on the absolute difference in read depth between the bulks."

```{r}
#One way to assess filtering thresholds and check the quality of our data is by plotting histograms of the read depths. We can get an idea of where to draw our thresholds

# total read depth histogram:

ggplot(data = Gprime_df) + 
    geom_histogram(aes(x = DP.HIGH + DP.LOW), bins = 50) + 
    xlim(0,500)

# total reference allele frequency:
ggplot(data = Gprime_df) +
    geom_histogram(aes(x = REF_FRQ), bins = 50)

# this one looks different than their example, their ref_frq was centered around 0.5... but maybe their data is really different. This looks like the vast majority of the SNPs are reference allele? Not sure exactly what that means...


# We can plot our per-bulk SNP-index to check if our data is good.We expect to find two small peaks on each end and most of the SNPs should be approximiately normally distributed arround 0.5 in an F2 population. Here is the HIGH bulk for example:

# for innubila sample
ggplot(data = Gprime_df) +
    geom_histogram(aes(x = SNPindex.HIGH, bins = 50))

# for vir sample
ggplot(data = Gprime_df) +
    geom_histogram(aes(x = SNPindex.LOW), bins = 50)

# this is not an F2 population (not even possible with our virus) so this is maybe why these don't look as expected 

```

#### Filtering SNPs

"Now that we have an idea about our read depth distribution we can filter out low confidence SNPS. In general we recommend filtering extremely low and high coverage SNPs, either in both bulks (`minTotalDepth/maxTotalDepth`) and/or in each bulk separately (`minSampleDepth`). We have the option to filter based on reference allele frequency (`refAlleleFreq`), this removes SNPs that for some reason are over- or under-represented in *BOTH* bulks. We can also filter SNPs that have large discrepancies in read depth between the bulks(i.e. one bulk has a depth of 500 and the other has 5). Such discrepancies can throw off the G statistic. We can also use the GATK GQ score (Genotype Quality) to filter out low confidence SNPs. If the `verbose` parameter is set to `TRUE` (default) the function will report the numbers of SNPs filtered in each step."

```{r}
# use the filtering snps 
# minimum depth should probably be 10? This is total so both samples, basically same as min sample depth
# max depth of 200? not sure 
# don't have GATK score so not filtering by that 
# min coverage per sample of 5 reads, seems pretty low but there are inn samples that I'm sure have few reads 
# for refAlleleFreq this is keeping all SNPs with a REF_FRQ between 0.05 and 0.95
# looks like if I use refAlleleFreq = something over .05 it will give a significant result for the Hampel method... 

Gprime_df_filt <-
    filterSNPs(
        SNPset = Gprime_df,
        refAlleleFreq = 0.05,
        minTotalDepth = 10,
        maxTotalDepth = 200, 
        minSampleDepth = 5,
        verbose = TRUE
    )

```

#### Running G prime analysis

```{r}

# use Hampel method for outlier filter 
Gprime_df_filt_prime <- runGprimeAnalysis(Gprime_df_filt,
    windowSize = 2000,
    outlierFilter = "Hampel",
    filterThreshold = 0.1)

# there are many errors with this... 


#Due to the fact that p-values are estimated from the null distribution of G', an important check is to see if the null distribution of G' values is close to log normally distributed. For this purpose we use the `plotGprimeDist` function, which plots the G' histograms of both raw and filtered G' sets alongside the log-normal null distribution (which is reported in the legend). We can also use this to test which filtering method (Hampel or DeltaSNP) estimates a more accurate null distribution. If you use the `"deltaSNP"` method plotting G' distributions with different filter thresholds might also help reveal a better G' null distribution. 

# Hampel method
plotGprimeDist(SNPset = Gprime_df_filt_prime, outlierFilter = "Hampel")
# deltaSNP method 
plotGprimeDist(SNPset = Gprime_df_filt_prime, outlierFilter = "deltaSNP", filterThreshold = 0.1)

# The hampel method looks most like a log normal distribution

# plot number of SNPs in a window
p1 <- plotQTLStats(SNPset = Gprime_df_filt_prime, var = "nSNPs")
p1

# plot the G' statistic
# try to plot the thereshold can pass the FDR (q) of 0.01.
p3 <- plotQTLStats(SNPset = Gprime_df_filt_prime, var = "Gprime", plotThreshold = TRUE, q = 0.01)

p3

# neg log ten pvalue
QTLplots <- plotQTLStats(
    SNPset = Gprime_df_filt_prime, 
    var = "negLog10Pval", 
    plotThreshold = TRUE, 
    q = 0.01, 
    )
QTLplots

```


```{r}
ggplot(Gprime_df_filt_prime, aes(x=POS, y=negLog10Pval))+
  geom_point()+
  theme_classic() + xlab("Position") + ylab("Negative log10 p-value")
```


#### Original Filtering and running 
```{r}
# filtering how I originally did it, which ended with no significant peaks 
# min sample depth 4
# for refAlleleFreq this is keeping all SNPs with a REF_FRQ between 0.001 and 0.999

Gprime_df_filt2 <-
    filterSNPs(
        SNPset = Gprime_df,
        refAlleleFreq = 0.001,
        minTotalDepth = 6,
        maxTotalDepth = 200, 
        minSampleDepth = 4,
        verbose = TRUE
    )

# uses deltaSNP method for outlier filter

Gprime_df_filt_prime2 <- runGprimeAnalysis(Gprime_df_filt2,
    windowSize = 2000,
    outlierFilter = "deltaSNP",
    filterThreshold = 0.1)

# Hampel method
plotGprimeDist(SNPset = Gprime_df_filt_prime2, outlierFilter = "Hampel")
# deltaSNP method 
plotGprimeDist(SNPset = Gprime_df_filt_prime2, outlierFilter = "deltaSNP", filterThreshold = 0.1)

# both plots look like log normal distribution

p4 <- plotQTLStats(SNPset = Gprime_df_filt_prime2, var = "Gprime", plotThreshold = TRUE, q = 0.01)
p4
```

#### original filter and running with hampel method
```{r}
# filtering how I originally did it, which ended with no significant peaks 
# min sample depth 4
# for refAlleleFreq this is keeping all SNPs with a REF_FRQ between 0.001 and 0.999

Gprime_df_filt2 <-
    filterSNPs(
        SNPset = Gprime_df,
        refAlleleFreq = 0.001,
        minTotalDepth = 6,
        maxTotalDepth = 200, 
        minSampleDepth = 4,
        verbose = TRUE
    )

# uses deltaSNP method for outlier filter

Gprime_df_filt_prime2 <- runGprimeAnalysis(Gprime_df_filt2,
    windowSize = 2000,
    outlierFilter = "Hampel",
    filterThreshold = 0.1)

# Hampel method
plotGprimeDist(SNPset = Gprime_df_filt_prime2, outlierFilter = "Hampel")
# deltaSNP method 
plotGprimeDist(SNPset = Gprime_df_filt_prime2, outlierFilter = "deltaSNP", filterThreshold = 0.1)

# both plots look like log normal distribution

p4 <- plotQTLStats(SNPset = Gprime_df_filt_prime2, var = "Gprime", plotThreshold = TRUE, q = 0.01)
p4
```

#### Try filtering out and singleton reads before filtering
```{r}
# remove rows where the alternate read is 1 for both
DiNV_SNPs_filt2 <- DiNV_SNPs[DiNV_SNPs$AD_ALT.inn != 1 & DiNV_SNPs$AD_ALT.vir != 1, ] 
# remove rows ehere reference read is 1 for both 
DiNV_SNPs_filt2 <- DiNV_SNPs_filt2[DiNV_SNPs_filt2$AD_REF.inn != 1 & DiNV_SNPs_filt2$AD_REF.vir != 1, ] 

# save this as a csv because the program doesn't seem to want it in R 
write.csv(DiNV_SNPs_filt2, file = "~/Desktop/KU/sequences/16Cq-DiNV-Test/freebayes/DiNV_SNPs_For_Gprime2.csv", row.names = FALSE)

# import from table
Gprime_df2 <- importFromTable(file = "~/Desktop/KU/sequences/16Cq-DiNV-Test/freebayes/DiNV_SNPs_For_Gprime2.csv",
                      highBulk = HighBulk,
                      lowBulk = LowBulk,
                      chromList = Chroms)

# check the hisograms 
# total read depth histogram:

ggplot(data = Gprime_df2) + 
    geom_histogram(aes(x = DP.HIGH + DP.LOW), bins = 50) + 
    xlim(0,500)

# total reference allele frequency:
ggplot(data = Gprime_df2) +
    geom_histogram(aes(x = REF_FRQ), bins = 50)

# this one looks different than their example, their ref_frq was centered around 0.5... but maybe their data is really different. This looks like the vast majority of the SNPs are reference allele? Not sure exactly what that means...


# We can plot our per-bulk SNP-index to check if our data is good.We expect to find two small peaks on each end and most of the SNPs should be approximiately normally distributed arround 0.5 in an F2 population. Here is the HIGH bulk for example:

# for innubila sample
ggplot(data = Gprime_df2) +
    geom_histogram(aes(x = SNPindex.HIGH, bins = 50))

# for vir sample
ggplot(data = Gprime_df2) +
    geom_histogram(aes(x = SNPindex.LOW), bins = 50)

# these generally look similar to the ones above

```

#### Filter and run Gprime analysis with this dataset
```{r}
prime_df_filt3 <-
    filterSNPs(
        SNPset = Gprime_df2,
        refAlleleFreq = 0.05,
        minTotalDepth = 10,
        maxTotalDepth = 200, 
        minSampleDepth = 5,
        verbose = TRUE
    )
```


```{r}
# uses Hampel method for outlier filter

Gprime_df_filt_prime3 <- runGprimeAnalysis(prime_df_filt3,
    windowSize = 2000,
    outlierFilter = "Hampel",
    filterThreshold = 0.1)

# Hampel method
plotGprimeDist(SNPset = Gprime_df_filt_prime3, outlierFilter = "Hampel")
# deltaSNP method 
plotGprimeDist(SNPset = Gprime_df_filt_prime3, outlierFilter = "deltaSNP", filterThreshold = 0.1)

# both plots look like log normal distribution

p4 <- plotQTLStats(SNPset = Gprime_df_filt_prime3, var = "Gprime", plotThreshold = TRUE, q = 0.01)
p4
```

#### Filter for less depth
```{r}
prime_df_filt4 <-
    filterSNPs(
        SNPset = Gprime_df2,
        refAlleleFreq = 0.05,
        minTotalDepth = 8,
        maxTotalDepth = 200, 
        minSampleDepth = 4,
        verbose = TRUE
    )
```


```{r}
# uses Hampel method for outlier filter

Gprime_df_filt_prime4 <- runGprimeAnalysis(prime_df_filt4,
    windowSize = 2000,
    outlierFilter = "Hampel",
    filterThreshold = 0.1)

# Hampel method
plotGprimeDist(SNPset = Gprime_df_filt_prime4, outlierFilter = "Hampel")
# deltaSNP method 
plotGprimeDist(SNPset = Gprime_df_filt_prime4, outlierFilter = "deltaSNP", filterThreshold = 0.1)

# both plots look like log normal distribution

p4 <- plotQTLStats(SNPset = Gprime_df_filt_prime4, var = "Gprime", plotThreshold = TRUE, q = 0.01)
p4
```

#### Do the SNPs filtered with original filtering (714) overlap with the 555 from the filtering after removing single read SNPs?

```{r}
# vector of positions of SNPs from original filtering, 714
positions_original <- Gprime_df_filt_prime$POS

# how many of these are in the 555 from the second filtering?
table(positions_original %in% prime_df_filt3$POS)
# ok so there are exactly 555 that match and it's the 159 that are lost, which are probably the singletons

# make a vector of the 159 that don't match
```

#### Do those 159 SNPs filtered out matter?

```{r}
# what are the 159 not shared snps
not_shared <- anti_join(Gprime_df_filt_prime, prime_df_filt3, by = "POS")
# make a vector of the positions
not_shared_pos <- not_shared$POS

# get dataset of all significant SNPs for original filtering (714)
QTL <- getSigRegions(SNPset = Gprime_df_filt_prime, alpha = 0.01)
# this is a list of 8, so it says there are 8 QTLs... 
# make these all into dataframes
QTL1 <- QTL[[1]]
QTL2 <- QTL[[2]]
QTL3 <- QTL[[3]]
QTL4 <- QTL[[4]]
QTL5 <- QTL[[5]]
QTL6 <- QTL[[6]]
QTL7 <- QTL[[7]]
QTL8 <- QTL[[8]]
# combine these all into 1 for now 

QTLS <- rbind(QTL1, QTL2, QTL3, QTL4, QTL5, QTL6, QTL7, QTL8)

# are there not shared snp positions in the QTLs of the originial filtering (714)
table(not_shared_pos %in% QTLS$POS)
# yes there are 34 positions (SNPs)

# what are those 34 SNPS?
not_shared_sig_snps <- not_shared_pos[not_shared_pos %in% QTLS$POS]
# make them into a df 
not_shared_sign_SNPS <- Gprime_df_filt_prime[Gprime_df_filt_prime$POS %in% not_shared_sig_snps,]
# look at the dataset 

# ok after looking at this dataset, I can tell that the filtering out of rows with 1 read was not what I should do. 
not_shared_sign_SNPS[5,2:10]
# see here one sample has all alternate and no reference, and the other sample has almost all ref and only 1 read for alternate. This is the exact kind of sample I want to keep 

# however there are also samples like this:
not_shared_sign_SNPS[7,2:10]
# where both samples are mostly ref 
# I am not sure why it kept this SNP or it came out as significant because the two samples basically are the same here 
# I need to figure out a way to remove these kind of SNPs, where one allele has only 1 read, but only if it basically is showing the same pattern as the other sample. 
```



```{r}

```







save filtered dataset from here
```{r}
#write.csv(Gprime_df_filt_prime4, "~/Desktop/KU/sequences/16Cq-DiNV-Test/freebayes/Gprime_df_filt.csv")

```

