---
title: "Analysis of pooled-continuous freebayes vcf file"
output: github_document
date: '2023-07-25'
---

### Read in vcf and subset out the allele counts from each sample
```{r}
#load library
#install.packages("vcfR")
library(vcfR)

#read in vcf
vcfR<-read.vcfR("~/Desktop/KU/sequences/16Cq-DiNV-Test/freebayes/continuous-DiNV.vcf")

#extract the number of reads supporting the reference allele for each sample at each SNP 
ref<-as.data.frame(extract.gt(vcfR, element = "RO"))
#extract the number of reads supporting the alternate allele for each sample at each SNP 
alt<-as.data.frame(extract.gt(vcfR, element = "AO"))

#subset out four columns that indicate in order: chromosome, position on the chromosome, reference allele, alternate allele(s)
df<-as.data.frame(vcfR@fix[,c(1,2,4,5)])
#add the read counts supporting the reference and alternate alleles for each of the two sequenced strains as their own columns
df$inn.ref<-ref$INN
df$inn.alt<-alt$INN
df$vir.ref<-ref$VIR
df$vir.alt<-alt$VIR

#check out the DF and make sure it looks right
head(df)
```

### Perform Fisher's exact test
```{r}
#for the first SNP, generate a dataframe that is in the right format for Fisher's exact test
df1<-data.frame(reference=as.numeric(c(df$inn.ref[1],df$vir.ref[1])),
                alt=as.numeric(c(df$inn.alt[1],df$vir.alt[1])),
                row.names = c("inn","vir"))

#perform a fisher's exact test
test1<-fisher.test(df1)
test1$p.value

#how can we scale this up to run on each SNP without copying and pasting this code 13,115 more times?
```










